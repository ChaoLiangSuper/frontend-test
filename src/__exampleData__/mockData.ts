export const mockCoreData = {
  id: 'XXXXXX-ID',
  number: 'XXXXXX-NUM',
  createdDate: '2017-07-10 15:14:01',
  lastUpdateDate: '2019-05-27 19:56:17',
  type: 'type',
  source: 'source',
  state: 'state',
  shortDescription: 'short description',
  application: 'application',
  assignee: 'assignee',
};

export const mockServiceData = {
  test1: 'test1',
  test2: 'test2',
  test3: 'test3',
};

export const mockCardData = {
  coreData: mockCoreData,
  serviceData: mockServiceData,
};
