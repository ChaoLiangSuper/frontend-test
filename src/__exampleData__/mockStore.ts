import { createStore } from 'redux';

export const mockStore = createStore(() => ({ totalPage: 0, pages: {} }));
