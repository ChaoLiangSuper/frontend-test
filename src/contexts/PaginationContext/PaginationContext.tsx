import _ from 'lodash';
import React, { createContext, useState } from 'react';
import { connect } from 'react-redux';
import { IPaginationContext, IStore } from '../../interfaces';

export const Context = createContext<IPaginationContext>({
  currentPage: 1,
  totalPage: 1,
  nextPage: _.noop,
  lastPage: _.noop,
});

export const CustomProvider = ({
  totalPage,
  children,
}: {
  totalPage: number;
  children: React.ReactNode;
}) => {
  const [state, setState] = useState<number>(1);

  const nextPage = () => {
    const next = state + 1;
    setState(next > totalPage ? totalPage : next);
  };

  const lastPage = () => {
    const last = state - 1;
    setState(last < 1 ? 1 : last);
  };

  return (
    <Context.Provider
      value={{ currentPage: state, totalPage, nextPage, lastPage }}
    >
      {children}
    </Context.Provider>
  );
};

export const PaginationProvider = connect(({ totalPage }: IStore) => ({
  totalPage,
}))(CustomProvider);

export const PaginationConsumer = Context.Consumer;
