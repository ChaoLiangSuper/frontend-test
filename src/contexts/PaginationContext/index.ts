export { PaginationProvider, PaginationConsumer } from './PaginationContext';
