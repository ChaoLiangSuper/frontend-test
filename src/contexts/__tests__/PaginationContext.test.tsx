import { act, renderHook } from '@testing-library/react-hooks';
import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { mockStore } from '../../__exampleData__/mockStore';
import {
  PaginationConsumer,
  PaginationProvider,
} from '../../contexts/PaginationContext';
import { CustomProvider } from '../../contexts/PaginationContext/PaginationContext';

describe('<PaginationContext /> component', () => {
  it('should successfully render without crash', () => {
    const wrapper = mount(
      <Provider store={mockStore}>
        <PaginationProvider>
          <PaginationConsumer>{() => null}</PaginationConsumer>
        </PaginationProvider>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default hook state', () => {
    const { result } = renderHook(() =>
      CustomProvider({ totalPage: 2, children: null })
    );

    expect(result.current.props.value.currentPage).toEqual(1);
    expect(result.current.props.value.totalPage).toEqual(2);

    act(() => {
      result.current.props.value.nextPage();
    });
    expect(result.current.props.value.currentPage).toEqual(2);

    act(() => {
      result.current.props.value.nextPage();
    });
    expect(result.current.props.value.currentPage).toEqual(2);

    act(() => {
      result.current.props.value.lastPage();
    });
    expect(result.current.props.value.currentPage).toEqual(1);

    act(() => {
      result.current.props.value.lastPage();
    });
    expect(result.current.props.value.currentPage).toEqual(1);
  });
});
