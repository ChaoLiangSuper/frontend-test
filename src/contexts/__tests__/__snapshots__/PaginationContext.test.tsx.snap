// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationContext /> component should successfully render without crash 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(CustomProvider)>
    <CustomProvider
      dispatch={[Function]}
      totalPage={0}
    />
  </Connect(CustomProvider)>
</Provider>
`;
